File: packages/node/src/streaming.ts
import sodium from "libsodium-wrappers";

export type StreamingKey = Uint8Array; // 32 bytes

function le64(n: number): Uint8Array {
  const b = new Uint8Array(8);
  let x = BigInt(n);
  for (let i = 0; i < 8; i++) {
    b[i] = Number(x & 0xffn);
    x >>= 8n;
  }
  return b;
}

export async function sealChunkXChaCha(
  key: StreamingKey,
  noncePrefix16: Uint8Array,
  seq: number,
  plaintext: Uint8Array,
  aad?: Uint8Array
): Promise<Uint8Array> {
  await sodium.ready;
  if (noncePrefix16.length !== 16) throw new Error("STREAM_NONCE_PREFIX_LEN");
  const nonce = new Uint8Array(24);
  nonce.set(noncePrefix16, 0);
  nonce.set(le64(seq), 16);
  return sodium.crypto_aead_xchacha20poly1305_ietf_encrypt(plaintext, aad ?? null, null, nonce, key);
}

export async function openChunkXChaCha(
  key: StreamingKey,
  noncePrefix16: Uint8Array,
  seq: number,
  ciphertext: Uint8Array,
  aad?: Uint8Array
): Promise<Uint8Array> {
  await sodium.ready;
  if (noncePrefix16.length !== 16) throw new Error("STREAM_NONCE_PREFIX_LEN");
  const nonce = new Uint8Array(24);
  nonce.set(noncePrefix16, 0);
  nonce.set(le64(seq), 16);
  return sodium.crypto_aead_xchacha20poly1305_ietf_decrypt(null, ciphertext, aad ?? null, nonce, key);
}

File: packages/python/x402_hpke/streaming.py
from __future__ import annotations
from typing import Optional
from nacl import bindings


def _le64(n: int) -> bytes:
    if n < 0:
        raise ValueError("SEQ_NEGATIVE")
    b = bytearray(8)
    x = n
    for i in range(8):
        b[i] = x & 0xFF
        x >>= 8
    return bytes(b)


def seal_chunk_xchacha(key: bytes, nonce_prefix16: bytes, seq: int, chunk: bytes, aad: Optional[bytes] = None) -> bytes:
    if len(key) != 32:
        raise ValueError("STREAM_KEY_LEN")
    if len(nonce_prefix16) != 16:
        raise ValueError("STREAM_NONCE_PREFIX_LEN")
    nonce = nonce_prefix16 + _le64(seq)
    return bindings.crypto_aead_xchacha20poly1305_ietf_encrypt(chunk, aad, nonce, key)


def open_chunk_xchacha(key: bytes, nonce_prefix16: bytes, seq: int, ciphertext: bytes, aad: Optional[bytes] = None) -> bytes:
    if len(key) != 32:
        raise ValueError("STREAM_KEY_LEN")
    if len(nonce_prefix16) != 16:
        raise ValueError("STREAM_NONCE_PREFIX_LEN")
    nonce = nonce_prefix16 + _le64(seq)
    return bindings.crypto_aead_xchacha20poly1305_ietf_decrypt(ciphertext, aad, nonce, key)

